File: packages/opencv_object_detection/android/src/main/java/com/grozziie/opencv_object_detection/YoloPostprocessor.java
package com.grozziie.opencv_object_detection;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/** Pure-Java YOLO decoding and class-aware non-maximum suppression. */
final class YoloPostprocessor {
    private YoloPostprocessor() {}

    static List<DetectionModels.Detection> decode(
            float[] output,
            int[] outputDimensions,
            List<String> labels,
            DetectionModels.LetterboxTransform transform,
            float confidenceThreshold,
            float iouThreshold,
            int maxResults
    ) {
        if (output.length == 0 || labels.isEmpty() || maxResults <= 0) {
            return Collections.emptyList();
        }
        int featureCount = labels.size() + 4;
        int[] dimensions = positiveDimensions(outputDimensions);
        int last = dimensions.length == 0 ? 0 : dimensions[dimensions.length - 1];
        int secondLast = dimensions.length < 2 ? 0 : dimensions[dimensions.length - 2];

        List<Candidate> candidates;
        if (secondLast == featureCount && last > 0) {
            candidates = decodeRaw(
                    output,
                    last,
                    featureCount,
                    true,
                    transform,
                    labels,
                    confidenceThreshold
            );
        } else if (last == featureCount && secondLast > 0) {
            candidates = decodeRaw(
                    output,
                    secondLast,
                    featureCount,
                    false,
                    transform,
                    labels,
                    confidenceThreshold
            );
        } else if (output.length % featureCount == 0) {
            candidates = decodeRaw(
                    output,
                    output.length / featureCount,
                    featureCount,
                    true,
                    transform,
                    labels,
                    confidenceThreshold
            );
        } else {
            return Collections.emptyList();
        }

        candidates.sort(Comparator.comparingDouble((Candidate value) -> value.confidence).reversed());
        List<Candidate> kept = classAwareNms(candidates, iouThreshold, maxResults);
        List<DetectionModels.Detection> results = new ArrayList<>(kept.size());
        for (Candidate candidate : kept) {
            results.add(new DetectionModels.Detection(
                    candidate.left,
                    candidate.top,
                    candidate.right,
                    candidate.bottom,
                    labels.get(candidate.classIndex),
                    candidate.classIndex,
                    candidate.confidence
            ));
        }
        return results;
    }

    static boolean hasSupportedLayout(int[] outputDimensions, int classCount) {
        int[] dimensions = positiveDimensions(outputDimensions);
        if (dimensions.length < 2) return false;
        int features = classCount + 4;
        int last = dimensions[dimensions.length - 1];
        int secondLast = dimensions[dimensions.length - 2];
        return last == features || secondLast == features;
    }

    private static List<Candidate> decodeRaw(
            float[] output,
            int anchorCount,
            int featureCount,
            boolean channelsFirst,
            DetectionModels.LetterboxTransform transform,
            List<String> labels,
            float confidenceThreshold
    ) {
        List<Candidate> candidates = new ArrayList<>();
        for (int anchor = 0; anchor < anchorCount; anchor++) {
            int classIndex = -1;
            float confidence = Float.NEGATIVE_INFINITY;
            for (int classOffset = 0; classOffset < labels.size(); classOffset++) {
                float score = value(
                        output,
                        anchor,
                        classOffset + 4,
                        anchorCount,
                        featureCount,
                        channelsFirst
                );
                if (Float.isFinite(score) && score > confidence) {
                    confidence = score;
                    classIndex = classOffset;
                }
            }
            if (classIndex < 0
                    || confidence < confidenceThreshold
                    || labels.get(classIndex).trim().equalsIgnoreCase("person")) {
                continue;
            }

            float centerX = value(output, anchor, 0, anchorCount, featureCount, channelsFirst);
            float centerY = value(output, anchor, 1, anchorCount, featureCount, channelsFirst);
            float width = value(output, anchor, 2, anchorCount, featureCount, channelsFirst);
            float height = value(output, anchor, 3, anchorCount, featureCount, channelsFirst);
            if (!Float.isFinite(centerX)
                    || !Float.isFinite(centerY)
                    || !Float.isFinite(width)
                    || !Float.isFinite(height)
                    || width <= 0f
                    || height <= 0f) {
                continue;
            }
            float maxCoordinate = Math.max(
                    Math.max(Math.abs(centerX), Math.abs(centerY)),
                    Math.max(Math.abs(width), Math.abs(height))
            );
            if (maxCoordinate <= 2f) {
                centerX *= transform.modelWidth;
                centerY *= transform.modelHeight;
                width *= transform.modelWidth;
                height *= transform.modelHeight;
            }
            Candidate candidate = modelBoxToCandidate(
                    centerX - width / 2f,
                    centerY - height / 2f,
                    centerX + width / 2f,
                    centerY + height / 2f,
                    transform,
                    classIndex,
                    confidence
            );
            if (candidate != null) candidates.add(candidate);
        }
        return candidates;
    }

    private static float value(
            float[] output,
            int anchor,
            int feature,
            int anchorCount,
            int featureCount,
            boolean channelsFirst
    ) {
        int index = channelsFirst
                ? feature * anchorCount + anchor
                : anchor * featureCount + feature;
        return index >= 0 && index < output.length ? output[index] : Float.NaN;
    }

    private static Candidate modelBoxToCandidate(
            float modelLeft,
            float modelTop,
            float modelRight,
            float modelBottom,
            DetectionModels.LetterboxTransform transform,
            int classIndex,
            float confidence
    ) {
        float left = clamp01(
                (modelLeft - transform.padX) / transform.scale / transform.sourceWidth
        );
        float top = clamp01(
                (modelTop - transform.padY) / transform.scale / transform.sourceHeight
        );
        float right = clamp01(
                (modelRight - transform.padX) / transform.scale / transform.sourceWidth
        );
        float bottom = clamp01(
                (modelBottom - transform.padY) / transform.scale / transform.sourceHeight
        );
        if (right <= left || bottom <= top) return null;
        return new Candidate(left, top, right, bottom, classIndex, confidence);
    }

    private static List<Candidate> classAwareNms(
            List<Candidate> sorted,
            float iouThreshold,
            int maxResults
    ) {
        List<Candidate> kept = new ArrayList<>(Math.min(sorted.size(), maxResults));
        for (Candidate candidate : sorted) {
            boolean overlaps = false;
            for (Candidate existing : kept) {
                if (existing.classIndex == candidate.classIndex
                        && intersectionOverUnion(existing, candidate) > iouThreshold) {
                    overlaps = true;
                    break;
                }
            }
            if (overlaps) continue;
            kept.add(candidate);
            if (kept.size() >= maxResults) break;
        }
        return kept;
    }

    private static float intersectionOverUnion(Candidate a, Candidate b) {
        float width = Math.max(0f, Math.min(a.right, b.right) - Math.max(a.left, b.left));
        float height = Math.max(0f, Math.min(a.bottom, b.bottom) - Math.max(a.top, b.top));
        float intersection = width * height;
        float aArea = (a.right - a.left) * (a.bottom - a.top);
        float bArea = (b.right - b.left) * (b.bottom - b.top);
        float union = aArea + bArea - intersection;
        return union > 0f ? intersection / union : 0f;
    }

    private static int[] positiveDimensions(int[] dimensions) {
        int count = 0;
        for (int dimension : dimensions) {
            if (dimension > 0) count++;
        }
        int[] result = new int[count];
        int target = 0;
        for (int dimension : dimensions) {
            if (dimension > 0) result[target++] = dimension;
        }
        return result;
    }

    private static float clamp01(float value) {
        return Math.max(0f, Math.min(1f, value));
    }

    private static final class Candidate {
        final float left;
        final float top;
        final float right;
        final float bottom;
        final int classIndex;
        final float confidence;

        Candidate(
                float left,
                float top,
                float right,
                float bottom,
                int classIndex,
                float confidence
        ) {
            this.left = left;
            this.top = top;
            this.right = right;
            this.bottom = bottom;
            this.classIndex = classIndex;
            this.confidence = confidence;
        }
    }
}

File: packages/native_object_detection/android/src/main/kotlin/com/grozziie/native_object_detection/YoloPostprocessor.kt
package com.grozziie.native_object_detection

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/** Pure-Kotlin YOLO decoding and class-aware non-maximum suppression. */
internal object YoloPostprocessor {
    private data class Candidate(
        val left: Float,
        val top: Float,
        val right: Float,
        val bottom: Float,
        val classIndex: Int,
        val confidence: Float,
    )

    fun decode(
        output: FloatArray,
        outputDimensions: IntArray,
        labels: List<String>,
        transform: LetterboxTransform,
        confidenceThreshold: Float,
        iouThreshold: Float,
        maxResults: Int,
    ): List<NativeDetection> {
        if (output.isEmpty() || labels.isEmpty() || maxResults <= 0) return emptyList()

        val features = labels.size + 4
        val dimensions = outputDimensions.filter { it > 0 }
        val last = dimensions.lastOrNull() ?: 0
        val secondLast = dimensions.getOrNull(dimensions.lastIndex - 1) ?: 0
        val candidates = when {
            secondLast > 0 && last == 6 && output.size >= secondLast * last ->
                decodeEndToEnd(
                    output,
                    rowCount = secondLast,
                    rowLength = last,
                    transform = transform,
                    labels = labels,
                    confidenceThreshold = confidenceThreshold,
                )
            secondLast == features && last > 0 -> decodeRaw(
                output,
                anchorCount = last,
                featureCount = secondLast,
                channelsFirst = true,
                transform = transform,
                labels = labels,
                confidenceThreshold = confidenceThreshold,
            )
            last == features && secondLast > 0 -> decodeRaw(
                output,
                anchorCount = secondLast,
                featureCount = last,
                channelsFirst = false,
                transform = transform,
                labels = labels,
                confidenceThreshold = confidenceThreshold,
            )
            output.size % features == 0 -> decodeRaw(
                output,
                anchorCount = output.size / features,
                featureCount = features,
                channelsFirst = true,
                transform = transform,
                labels = labels,
                confidenceThreshold = confidenceThreshold,
            )
            else -> emptyList()
        }

        val kept = classAwareNms(
            candidates.sortedByDescending { it.confidence },
            iouThreshold,
            maxResults,
        )
        return kept.map { candidate ->
            NativeDetection(
                left = candidate.left,
                top = candidate.top,
                right = candidate.right,
                bottom = candidate.bottom,
                label = labels.getOrElse(candidate.classIndex) {
                    "class ${candidate.classIndex}"
                },
                classIndex = candidate.classIndex,
                confidence = candidate.confidence,
            )
        }
    }

    private fun decodeRaw(
        output: FloatArray,
        anchorCount: Int,
        featureCount: Int,
        channelsFirst: Boolean,
        transform: LetterboxTransform,
        labels: List<String>,
        confidenceThreshold: Float,
    ): List<Candidate> {
        val candidates = ArrayList<Candidate>()
        fun value(anchor: Int, feature: Int): Float {
            val index = if (channelsFirst) {
                feature * anchorCount + anchor
            } else {
                anchor * featureCount + feature
            }
            return output.getOrElse(index) { Float.NaN }
        }

        for (anchor in 0 until anchorCount) {
            var classIndex = -1
            var confidence = Float.NEGATIVE_INFINITY
            for (classOffset in labels.indices) {
                val score = value(anchor, classOffset + 4)
                if (score.isFinite() && score > confidence) {
                    confidence = score
                    classIndex = classOffset
                }
            }
            if (classIndex < 0 ||
                confidence < confidenceThreshold ||
                isPerson(classIndex, labels)
            ) {
                continue
            }

            var centerX = value(anchor, 0)
            var centerY = value(anchor, 1)
            var width = value(anchor, 2)
            var height = value(anchor, 3)
            if (!centerX.isFinite() ||
                !centerY.isFinite() ||
                !width.isFinite() ||
                !height.isFinite() ||
                width <= 0f ||
                height <= 0f
            ) {
                continue
            }
            val coordinatesAreNormalized = max(
                max(abs(centerX), abs(centerY)),
                max(abs(width), abs(height)),
            ) <= 2f
            if (coordinatesAreNormalized) {
                centerX *= transform.modelWidth
                centerY *= transform.modelHeight
                width *= transform.modelWidth
                height *= transform.modelHeight
            }

            modelBoxToCandidate(
                centerX - width / 2f,
                centerY - height / 2f,
                centerX + width / 2f,
                centerY + height / 2f,
                transform,
                classIndex,
                confidence,
            )?.let(candidates::add)
        }
        return candidates
    }

    private fun decodeEndToEnd(
        output: FloatArray,
        rowCount: Int,
        rowLength: Int,
        transform: LetterboxTransform,
        labels: List<String>,
        confidenceThreshold: Float,
    ): List<Candidate> {
        val candidates = ArrayList<Candidate>()
        for (row in 0 until rowCount) {
            val base = row * rowLength
            val confidence = output.getOrElse(base + 4) { Float.NaN }
            val classIndex = output.getOrElse(base + 5) { Float.NaN }.toInt()
            if (!confidence.isFinite() ||
                confidence < confidenceThreshold ||
                classIndex !in labels.indices ||
                isPerson(classIndex, labels)
            ) {
                continue
            }

            var left = output.getOrElse(base) { Float.NaN }
            var top = output.getOrElse(base + 1) { Float.NaN }
            var right = output.getOrElse(base + 2) { Float.NaN }
            var bottom = output.getOrElse(base + 3) { Float.NaN }
            if (!left.isFinite() ||
                !top.isFinite() ||
                !right.isFinite() ||
                !bottom.isFinite()
            ) {
                continue
            }
            val coordinatesAreNormalized = max(
                max(abs(left), abs(top)),
                max(abs(right), abs(bottom)),
            ) <= 2f
            if (coordinatesAreNormalized) {
                left *= transform.modelWidth
                right *= transform.modelWidth
                top *= transform.modelHeight
                bottom *= transform.modelHeight
            }
            modelBoxToCandidate(
                left,
                top,
                right,
                bottom,
                transform,
                classIndex,
                confidence,
            )?.let(candidates::add)
        }
        return candidates
    }

    private fun modelBoxToCandidate(
        modelLeft: Float,
        modelTop: Float,
        modelRight: Float,
        modelBottom: Float,
        transform: LetterboxTransform,
        classIndex: Int,
        confidence: Float,
    ): Candidate? {
        val left = ((modelLeft - transform.padX) / transform.scale /
            transform.sourceWidth).coerceIn(0f, 1f)
        val top = ((modelTop - transform.padY) / transform.scale /
            transform.sourceHeight).coerceIn(0f, 1f)
        val right = ((modelRight - transform.padX) / transform.scale /
            transform.sourceWidth).coerceIn(0f, 1f)
        val bottom = ((modelBottom - transform.padY) / transform.scale /
            transform.sourceHeight).coerceIn(0f, 1f)
        if (right <= left || bottom <= top) return null
        return Candidate(left, top, right, bottom, classIndex, confidence)
    }

    private fun classAwareNms(
        sorted: List<Candidate>,
        iouThreshold: Float,
        maxResults: Int,
    ): List<Candidate> {
        val kept = ArrayList<Candidate>(min(maxResults, sorted.size))
        for (candidate in sorted) {
            if (kept.any {
                    it.classIndex == candidate.classIndex &&
                        intersectionOverUnion(it, candidate) > iouThreshold
                }
            ) {
                continue
            }
            kept.add(candidate)
            if (kept.size >= maxResults) break
        }
        return kept
    }

    private fun intersectionOverUnion(a: Candidate, b: Candidate): Float {
        val intersectionWidth = max(0f, min(a.right, b.right) - max(a.left, b.left))
        val intersectionHeight = max(0f, min(a.bottom, b.bottom) - max(a.top, b.top))
        val intersection = intersectionWidth * intersectionHeight
        val aArea = (a.right - a.left) * (a.bottom - a.top)
        val bArea = (b.right - b.left) * (b.bottom - b.top)
        val union = aArea + bArea - intersection
        return if (union > 0f) intersection / union else 0f
    }

    private fun isPerson(classIndex: Int, labels: List<String>): Boolean {
        return labels.getOrNull(classIndex)?.trim()?.equals("person", ignoreCase = true) == true
    }
}
